### PR TITLE
feat(compact): add FieldListaggAgg aggregate function

### PR DESCRIPTION
**Pull request by Zouxxyy**
## Summary

Add `listagg` aggregate function for MergeTree compaction. Concatenates
string values with a configurable delimiter, with optional distinct mode
to deduplicate tokens.

## Test plan

Unit tests cover null handling, empty strings, custom delimiters,
distinct dedup, and type validation.

🤖 Generated with [Claude Code]

### CODE REVIEW

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024-present Alibaba Inc.
+ *
```
Comment (lxy-9602) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h:3]: For new file, please update the copyright year to 2026.

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+static CoreOptions CreateOptions(const std::map<std::string, std::string>& opts = {}) {
+    auto result = CoreOptions::FromMap(opts);
+    if (!result.ok()) {
+        ADD_FAILURE() << result.status().ToString();
+        return CoreOptions();
+    }
+    return std::move(result.value());
+}
+
```
Comment (lxy-9602) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp:42]: To better organize the tests, consider creating a `FieldListaggAggTest` fixture class. Moving shared utilities there and using `TEST_F`. Also you can use `EXPECT_OK_AND_ASSIGN(auto core_options, CoreOptions::FromMap(opts));`

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+TEST(FieldListaggAggTest, TestDistinct) {
+    auto options = CreateOptions({{"fields.f.aggregate-function", "listagg"},
+                                  {"fields.f.list-agg-delimiter", ";"},
+                                  {"fields.f.distinct", "true"}});
+    ASSERT_OK_AND_ASSIGN(std::unique_ptr<FieldListaggAgg> field_listagg_agg,
+                         FieldListaggAgg::Create(arrow::utf8(), options, "f"));
+
+    // accumulator="a;b", input="b;c" -> result="a;b;c" (deduplicate "b")
+    auto agg_ret = field_listagg_agg->Agg(std::string_view("a;b"), std::string_view("b;c"));
+    ASSERT_EQ(DataDefine::GetVariantValue<std::string_view>(agg_ret), "a;b;c");
```
Comment (lxy-9602) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp]: Could we add a test case to verify that `distinct=false` actually preserves duplicate values? (like `TestDistinct`)

File: include/paimon/defs.h [hunk 4]
```diff
@@ -79,6 +79,10 @@ struct PAIMON_EXPORT Options {
     static const char DEFAULT_AGG_FUNCTION[];
     /// IGNORE_RETRACT is "ignore-retract"
     static const char IGNORE_RETRACT[];
+    /// DISTINCT is "distinct"
+    static const char DISTINCT[];
```
Comment (lxy-9602) [inline include/paimon/defs.h:84]: Could you please add the default value and a usage example in the comment?  
For example: `distinct=false` by default, and how to set it via config like `fields.f.distinct=false`.

File: src/paimon/core/core_options.cpp [hunk 5]
```diff
@@ -1004,6 +1004,24 @@ Result<bool> CoreOptions::FieldAggIgnoreRetract(const std::string& field_name) c
...
+Result<bool> CoreOptions::FieldCollectAggDistinct(const std::string& field_name) const {
+    ConfigParser parser(impl_->raw_options);
+    bool distinct = false;
+    std::string key = std::string(Options::FIELDS_PREFIX) + "." + field_name + "." +
+                      std::string(Options::DISTINCT);
+    PAIMON_RETURN_NOT_OK(parser.Parse<bool>(key, &distinct));
+    return distinct;
+}
+
 bool CoreOptions::DeletionVectorsEnabled() const {
```
Comment (lxy-9602) [inline src/paimon/core/core_options.cpp:1135]: Could you please add test for the above two func in `CoreOptionsTest, TestDefaultValue` and `CoreOptionsTest, TestFromMap`

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        if (field_type->id() != arrow::Type::type::STRING) {
+            return Status::Invalid(
+                fmt::format("invalid field type {} for {}, supposed to be string",
+                            field_type->ToString(), NAME));
+        }
+        PAIMON_ASSIGN_OR_RAISE(std::string delimiter, options.FieldListAggDelimiter(field_name));
+        PAIMON_ASSIGN_OR_RAISE(bool distinct, options.FieldCollectAggDistinct(field_name));
+        if (distinct && delimiter.empty()) {
+            return Status::Invalid(
+                fmt::format("invalid empty delimiter for {} when distinct is enabled", NAME));
```
Comment (lxy-9602) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h]: Is there a corresponding check in Java for `distinct && delimiter.empty()`?  
Java's `splitByWholeSeparatorPreserveAllTokens` seems to allow empty delimiters — should we align?

**Comment by lxy-9602** [comment]
Great work! Thanks for enhancing the aggregation functionality!

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+/// listagg aggregate a field of a row.
+/// Concatenates string values with a delimiter.
+class FieldListaggAgg : public FieldAggregator {
+ public:
+    static constexpr char NAME[] = "listagg";
+
+    static Result<std::unique_ptr<FieldListaggAgg>> Create(
+        const std::shared_ptr<arrow::DataType>& field_type, const CoreOptions& options,
+        const std::string& field_name) {
+        if (field_type->id() != arrow::Type::type::STRING) {
```
Comment (lszskye) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h:41]: please print field_name here

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            new_result.append(acc_str);
+            new_result.append(delimiter_);
+            new_result.append(in_str);
+            result_ = std::move(new_result);
+        }
+        return std::string_view(result_);
+    }
+
+ private:
+    std::string AggDistinctImpl(std::string_view acc_str, std::string_view in_str) {
```
Comment (lszskye) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h]: std::string AggDistinctImpl(std::string_view acc_str, std::string_view in_str) const

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2024-present Alibaba Inc.
```
Comment (lszskye) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg_test.cpp]: Copyright 2026-present Alibaba Inc.

File: src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            return Status::Invalid(
+                fmt::format("invalid field type {} for field '{}' of {}, supposed to be string",
+                            field_type->ToString(), field_name, NAME));
+        }
+        PAIMON_ASSIGN_OR_RAISE(std::string delimiter, options.FieldListAggDelimiter(field_name));
+        PAIMON_ASSIGN_OR_RAISE(bool distinct, options.FieldCollectAggDistinct(field_name));
+        // When delimiter is empty, fall back to whitespace split
+        if (delimiter.empty()) {
+            delimiter = " ";
+        }
```
Comment (lxy-9602) [inline src/paimon/core/mergetree/compact/aggregate/field_listagg_agg.h:51]: It looks like the Java logic is: if `distinct` is true, it applies a fallback mechanism (e.g. deduplication), otherwise it directly concatenates the values.

**Comment by lxy-9602** [comment]
By the way, your change helped us identify a bug in the Java implementation:  
In `distinct` mode, it uses `contains()` to check for duplicates, which can cause false positives due to substring matching.  
For example, `agg("ab", "a")` would incorrectly treat `"a"` as duplicate even though it's not.  

We're preparing a PR today to fix this on the Java side — thanks for catching this discrepancy!

**Review by lxy-9602 — APPROVED** [review]
+1

**Comment by Zouxxyy** [comment]
@lxy-9602 @lszskye Thanks for the review!
